api/RestfulService.php

```diff
@@ -164,25 +243,38 @@ protected function constructURL(){
 	 *                                   be 500
 	 */
 	public function request($subURL = '', $method = "GET", $data = null, $headers = null, $curlOptions = array()) {
-
+		//normalise varliables here
 		$url = $this->getAbsoluteRequestURL($subURL);
 		$method = strtoupper($method);
+		$headers = array_merge((array)$this->customHeaders, (array)$headers);
+		$curlOptions = $curlOptions + (array)$this->config()->default_curl_options;
+		$cacheExpire = $this->cache_expire;
+		$canCache = in_array($method, $this->config()->cachable_methods);
 
-		assert(in_array($method, array('GET','POST','PUT','DELETE','HEAD','OPTIONS')));
+		if (!in_array($method, array('GET', 'POST', 'PUT', 'PATCH', 'DELETE', 'HEAD', 'OPTIONS'))) {
+			throw new InvalidArgumentException('Invalid method "' . var_export($method, true) . '" passed to request');
+		}
 
-		$cache_path = $this->getCachePath(array(
-			$url,
-			$method,
-			$data,
-			array_merge((array)$this->customHeaders, (array)$headers),
-			$curlOptions + (array)$this->config()->default_curl_options,
-			$this->getBasicAuthString()
-		));
+		//if we can cache, calculate the cache path
+		if ($canCache) {
+			$cache_path = $this->getCachePath(array(
+				$url,
+				$method,
+				$data,
+				$headers,
+				$curlOptions,
+				$this->getBasicAuthString()
+			));
+		}
+		// otherwise set the cache expiry to 0
+		else {
+			$cacheExpire = 0;
+		}
 
 		// Check for unexpired cached feed (unless flush is set)
 		//assume any cache_expire that is 0 or less means that we dont want to
 		// cache
-		if($this->cache_expire > 0 && !self::$flush
+		if ($canCache && $this->cache_expire > 0 && !self::$flush
 				&& @file_exists($cache_path)
```

Comment:
This point was made previously and some of the work has been done here: https://github.com/silverstripe/silverstripe-framework/pull/3471

It wasn't something I was focussing on when I initially made these improvements and this is just a rebase of old work.

My view has shifted to getting people to use guzzle over this and adding a RestfulService poly-fill for those that want legacy support.

Core only uses RestfulService once in Oembed.php and that could be done with a raw curl request instead (as it's so trivial).

I'd therefore opt to remove any kind of client from core and users can require guzzle as needed